File: examples/example-java/services/intermediary-service/src/main/java/example/services/intermediary/IntermediaryService.java
package example.services.intermediary;

import example.services.balance.BalanceServiceGrpc;
import example.services.balance.GetBalanceRequest2;
import io.grpc.stub.StreamObserver;

public class IntermediaryService extends IntermediaryServiceGrpc.IntermediaryServiceImplBase {
    private final BalanceServiceGrpc.BalanceServiceBlockingStub balanceServiceClient;

    public IntermediaryService(BalanceServiceGrpc.BalanceServiceBlockingStub balanceServiceClient) {
        this.balanceServiceClient = balanceServiceClient;
    }

    @Override
    public void getBalance(GetBalanceRequest1 request, StreamObserver<GetBalanceResponse1> responseObserver) {
        var balanceResponse = balanceServiceClient.getBalance(
                GetBalanceRequest2.newBuilder()
                        .setUserId(request.getUserId())
                        .build()
        );

        responseObserver.onNext(GetBalanceResponse1.newBuilder()
                .setStatus(balanceResponse.getStatus())
                .setBalance(balanceResponse.getBalance()).build());
        responseObserver.onCompleted();
    }
}

File: examples/example-kotlin/services/intermediary-service/src/main/kotlin/IntermediaryService.kt
package example.services.intermediary

import example.services.balance.BalanceServiceGrpcKt.BalanceServiceCoroutineStub
import example.services.balance.getBalanceRequest2
import example.services.intermediary.IntermediaryServiceGrpcKt.IntermediaryServiceCoroutineImplBase

class IntermediaryService(private val balanceServiceClient: BalanceServiceCoroutineStub) :
    IntermediaryServiceCoroutineImplBase() {
    override suspend fun getBalance(request: GetBalanceRequest1): GetBalanceResponse1 {
        val balanceResponse = balanceServiceClient.getBalance(getBalanceRequest2 {
            userId = request.userId
        })

        return getBalanceResponse1 {
            status = balanceResponse.status
            balance = balanceResponse.balance
        }
    }
}
